README.md
# HTML Wrapper
<a href="http://flutter-seo.web.app"><img src="https://flutter-seo.web.app/assets/assets/wrapp.png"></img></a>

The `html_wrapper` package provides a simple solution for optimizing Search Engine Optimization (SEO) in Flutter applications by converting essential widgets into real HTML tags. By rendering Flutter widgets as true HTML elements, search engines can better understand the content of your app, leading to improved visibility and search rankings.

## Features

- Convert Flutter widgets into HTML tags to improve SEO.
- Preserve the functionality and styling of Flutter widgets while generating SEO-friendly HTML output.
- Customize HTML attributes and tags to align with SEO best practices.

## Getting Started

#### Installation

Click here: <a href="https://pub.dev/packages/html_wrapper">Pub dev</a>

Add the following dependency to your pubspec.yaml file:

```
dependencies:
    html_wrapper: ^0.0.4
```

Then, run `flutter pub get` to fetch the package.

## Usage

- import the html_wrapper package: ```import 'package:html_wrapper/html_wrapper.dart' as wrapp;```

The library is easy to use, the names and behaviors of the widgets remain the same as classic Widget one exclusivity in each widget is the implementation of `href` key that allow to click on it and redirect in another url using `<a>`tag.

#### Text Widget

the novelty of the Text widget compared to the real one is the addition of the tag key allowing to choose in which html tag we want our text to be between h1, span, strong etc...

```
wrapp.Text(
    'Hello Seo!',
    tag: wrapp.Tag.h1,
    style: TextStyle(
    fontFamily: 'arial',
    fontSize: MediaQuery.of(context).size.width / 6,
    color: Colors.white,
    fontWeight: FontWeight.w900),
),
```

#### Image Widget

all the parameters of an image are processed such as size, fit and urls or images from a folder are correctly display.

```
wrapp.Image(
    height: 200,
    width: 200,
    borderRadius: BorderRadius.circular(200),
    href: 'https://github.com/Antoinegtir',
    src: 'https://avatars.githubusercontent.com/u/114834504?v=4',
    fit: BoxFit.cover,
),
```

#### Robot Widget

In this example, we have a simple layout that displays the text "Hello, world!" in the center of the screen. Additionally, we include the Robots widget to set the `<meta>` tag for search engine crawlers.

The Robots widget is responsible for adding the `<meta>` tag with the specified content attribute value to the HTML head section. In this case, we set the content to 'noindex, nofollow', indicating that search engines should not index or follow the page.

```
Column(
    mainAxisAlignment: MainAxisAlignment.center,
    children: [
        Text(
        'Hello, world!',
        style: TextStyle(fontSize: 24),
        ),
        SizedBox(height: 16),
        Robots(content: 'noindex, nofollow'),
    ],
),
```

The project is only at the beginning, there are still a lot of tags to implement and optimization do not hesitate to read the contributing.md if you want to contribute to the project ❤️

CHANGELOG.md
## 0.0.4

* Implement Text, Image, Meta Widget

## 0.0.1

* TODO: Describe initial release.


lib/srcs/text.dart
// ignore_for_file: public_member_api_docs, prefer_single_quotes, avoid_redundant_argument_values, avoid_positional_boolean_parameters, cast_nullable_to_non_nullable, avoid_web_libraries_in_flutter, unsafe_html, unnecessary_cast

import 'dart:html' as html;
import 'package:flutter/material.dart';

class Tag {
  static const String p = "p";
  static const String u = "u";
  static const String s = "s";
  static const String q = "q";
  static const String h1 = "h1";
  static const String h2 = "h2";
  static const String h3 = "h3";
  static const String h4 = "h4";
  static const String h5 = "h5";
  static const String h6 = "h6";
  static const String em = "em";
  static const String sub = "sub";
  static const String sup = "sup";
  static const String del = "del";
  static const String ins = "ins";
  static const String pre = "pre";
  static const String big = "big";
  static const String abbr = "abbr";
  static const String span = "span";
  static const String code = "code";
  static const String small = "small";
  static const String strong = "strong";
  static const String mark = "mark";
  static const String address = "address";
  static const String blockquote = "blockquote";
}

class Text extends SingleChildRenderObjectWidget {
  const Text(
    this.text, {
    Key? key,
    this.style,
    this.strutStyle,
    this.textAlign,
    this.textDirection,
    this.locale,
    this.softWrap,
    this.overflow,
    this.textScaleFactor,
    this.maxLines,
    this.semanticsLabel,
    this.textWidthBasis,
    this.textHeightBehavior,
    this.selectionColor,
    required this.tag,
    this.href,
  }) : super(key: key);
  final String text;
  final TextStyle? style;
  final StrutStyle? strutStyle;
  final TextAlign? textAlign;
  final TextDirection? textDirection;
  final Locale? locale;
  final bool? softWrap;
  final TextOverflow? overflow;
  final double? textScaleFactor;
  final int? maxLines;
  final String? semanticsLabel;
  final TextWidthBasis? textWidthBasis;
  final TextHeightBehavior? textHeightBehavior;
  final Color? selectionColor;
  final String tag;
  final String? href;

  @override
  RenderObject createRenderObject(BuildContext context) {
    return RenderTextToHtml(
      text,
      style,
      strutStyle,
      textAlign,
      textDirection,
      locale,
      softWrap,
      overflow,
      textScaleFactor,
      maxLines,
      semanticsLabel,
      textWidthBasis,
      textHeightBehavior,
      selectionColor,
      tag,
      href,
    );
  }
}

class RenderTextToHtml extends RenderBox {
  RenderTextToHtml(
    this.text,
    this.style,
    this.strutStyle,
    this.textAlign,
    this.textDirection,
    this.locale,
    this.softWrap,
    this.overflow,
    this.textScaleFactor,
    this.maxLines,
    this.semanticsLabel,
    this.textWidthBasis,
    this.textHeightBehavior,
    this.selectionColor,
    this.tag,
    this.href,
  ) {
    if (tag == null) {
      element = html.Element.tag("span");
    }
    if (tag != null) {
      element = html.Element.tag(tag as String);
    }
  }
  final String? href;
  final String text;
  final TextStyle? style;
  final StrutStyle? strutStyle;
  final TextAlign? textAlign;
  final TextDirection? textDirection;
  final Locale? locale;
  final bool? softWrap;
  final TextOverflow? overflow;
  final double? textScaleFactor;
  final int? maxLines;
  final String? semanticsLabel;
  final TextWidthBasis? textWidthBasis;
  final TextHeightBehavior? textHeightBehavior;
  final Color? selectionColor;
  final String? tag;
  html.Element element = html.SpanElement();

  @override
  void performLayout() {
    element.text = text;
    element.style.color = _convertColorToCss(style?.color);
    element.style.fontSize = '${style?.fontSize}px';
    element.style.fontFamily = style?.fontFamily;
    element.style.fontWeight = _convertFontWeightToCss(style?.fontWeight);

    final TextPainter textPainter = TextPainter(
      text: TextSpan(text: text, style: style),
      textDirection: textDirection ?? TextDirection.ltr,
      textAlign: textAlign ?? TextAlign.start,
      textScaleFactor: textScaleFactor ?? 1.0,
      maxLines: maxLines ?? 1,
      strutStyle: strutStyle,
      ellipsis: overflow == TextOverflow.ellipsis ? '...' : null,
      locale: locale,
      textWidthBasis: textWidthBasis ?? TextWidthBasis.parent,
      textHeightBehavior: textHeightBehavior,
    );
    textPainter.layout(minWidth: 0, maxWidth: double.infinity);

    final bool wrapText = softWrap ?? false;
    final double maxWidth = constraints.maxWidth;
    if (wrapText && textPainter.width > maxWidth) {
      textPainter.layout(maxWidth: maxWidth);
    }

    final Size desiredSize = Size(
      textPainter.width,
      textPainter.height,
    );

    size = constraints.constrain(desiredSize);
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    final html.DivElement div = html.DivElement()
      ..style.position = 'absolute'
      ..style.left = '${offset.dx}px'
      ..style.top = '${offset.dy}px'
      ..style.width = '${size.width + style!.fontSize!.toInt()}px'
      ..style.height = '${size.height}px';

    final html.Element element = (href !=
            null)
        ? html
            .AnchorElement()
        : html
            .SpanElement();

    element.text = text;
    element.style.color = _convertColorToCss(style?.color);
    element.style.fontSize = '${style?.fontSize}px';
    element.style.fontFamily = style?.fontFamily;
    element.style.fontWeight = _convertFontWeightToCss(style?.fontWeight);

    if (element is html.AnchorElement && href != null) {
      (element as html.AnchorElement).href =
          href;
    }

    div.append(element);

    html.document.body?.append(div);
  }

  String? _convertColorToCss(Color? color) {
    if (color != null) {
      return 'rgb(${color.red}, ${color.green}, ${color.blue})';
    }
    return 'rgb(0,0,0)';
  }

  String? _convertFontWeightToCss(FontWeight? fontWeight) {
    if (fontWeight != null) {
      if (fontWeight == FontWeight.bold) {
        return 'bold';
      }
      if (fontWeight == FontWeight.normal) {
        return 'normal';
      }
      if (fontWeight == FontWeight.w100) {
        return '100';
      }
      if (fontWeight == FontWeight.w200) {
        return '200';
      }
      if (fontWeight == FontWeight.w300) {
        return '300';
      }
      if (fontWeight == FontWeight.w400) {
        return '400';
      }
      if (fontWeight == FontWeight.w500) {
        return '500';
      }
      if (fontWeight == FontWeight.w600) {
        return '600';
      }
      if (fontWeight == FontWeight.w700) {
        return '700';
      }
      if (fontWeight == FontWeight.w800) {
        return '800';
      }
      if (fontWeight == FontWeight.w900) {
        return '900';
      }
    }
    return null;
  }
}


lib/srcs/image.dart
// ignore_for_file: public_member_api_docs, avoid_web_libraries_in_flutter, unnecessary_cast, unsafe_html

import 'dart:html' as html;
import 'package:flutter/material.dart';

class Image extends SingleChildRenderObjectWidget { // Nouvel argument ajouté

  const Image({
    required this.src,
    this.width,
    this.height,
    this.fit = BoxFit.contain,
    Key? key,
    this.href,
    this.borderRadius, // Nouvel argument ajouté
  }) : super(key: key);
  final String src;
  final double? width;
  final double? height;
  final BoxFit fit;
  final String? href;
  final BorderRadius? borderRadius;

  @override
  RenderObject createRenderObject(BuildContext context) {
    return RenderImageToHtml(
      src,
      width,
      height,
      fit,
      href: href,
      borderRadius:
          borderRadius, // Passage de l'argument borderRadius au constructeur de RenderImageToHtml
    );
  }
}

class RenderImageToHtml extends RenderBox {

  RenderImageToHtml(
    this.src,
    this.width,
    this.height,
    this.fit, {
    this.href,
    this.borderRadius, // Nouvel argument ajouté
  });
  final String src;
  final double? width;
  final double? height;
  final BoxFit fit;
  final String? href;
  final BorderRadius? borderRadius; // Nouvel argument ajouté

  html.ImageElement? imageElement;

  @override
  void performLayout() {
    final double desiredWidth = width ?? constraints.maxWidth;
    final double desiredHeight = height ?? constraints.maxHeight;

    // Create a new HTML image element
    imageElement = html.ImageElement();

    // Set the loading listener to update the layout once the image is loaded

    // Set the image dimensions
    imageElement!.src = src;
    imageElement!.width = desiredWidth.toInt();
    imageElement!.height = desiredHeight.toInt();
    final String objectFit = _convertBoxFitToCss(fit);
    imageElement!.style.objectFit = objectFit;

    if (borderRadius != null) {
      final String borderRadiusCss = _convertBorderRadiusToCss(borderRadius!);
      imageElement!.style.borderRadius = borderRadiusCss;
    }

    // Update the size constraints
    size = constraints.constrain(Size(desiredWidth, desiredHeight));
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    if (imageElement != null) {
      final html.DivElement div = html.DivElement()
        ..style.position = 'absolute'
        ..style.left = '${offset.dx}px'
        ..style.top = '${offset.dy}px'
        ..style.width = '${size.width}px'
        ..style.height = '${size.height}px';

      final html.Element contentElement =
          (href != null) ? html.AnchorElement() : html.SpanElement();

      contentElement.style.display = 'block';
      contentElement.style.width = '100%';
      contentElement.style.height = '100%';
      contentElement.append(imageElement!);

      if (contentElement is html.AnchorElement && href != null) {
        (contentElement as html.AnchorElement).href = href;
      }

      div.append(contentElement);

      html.document.body?.append(div);
    }
  }

  String _convertBoxFitToCss(BoxFit fit) {
    switch (fit) {
      case BoxFit.fill:
        return 'fill';
      case BoxFit.contain:
        return 'contain';
      case BoxFit.cover:
        return 'cover';
      case BoxFit.fitWidth:
        return 'fit-width';
      case BoxFit.fitHeight:
        return 'fit-height';
      case BoxFit.none:
        return 'none';
      case BoxFit.scaleDown:
        return 'scale-down';
      default:
        return 'contain';
    }
  }

  String _convertBorderRadiusToCss(BorderRadius borderRadius) {
    final Radius topLeft = borderRadius.topLeft;
    final Radius topRight = borderRadius.topRight;
    final Radius bottomLeft = borderRadius.bottomLeft;
    final Radius bottomRight = borderRadius.bottomRight;

    return '${_convertRadiusToCss(topLeft)} '
        '${_convertRadiusToCss(topRight)} '
        '${_convertRadiusToCss(bottomRight)} '
        '${_convertRadiusToCss(bottomLeft)}';
  }

  String _convertRadiusToCss(Radius radius) {
    return '${radius.x}px';
  }
}


example/lib/main.dart
// ignore_for_file: avoid_web_libraries_in_flutter
import 'package:flutter/material.dart';
import 'package:html_wrapper/html_wrapper.dart' as wrapp;
import 'package:url_strategy/url_strategy.dart';

void main() {
  setPathUrlStrategy();
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        debugShowCheckedModeBanner: false,
        title: 'Flutter to HTML Wrapper',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: Scaffold(
            backgroundColor: Colors.black,
            extendBodyBehindAppBar: true,
            appBar: AppBar(
              backgroundColor: Colors.black,
            ),
            body: const SeoPage()));
  }
}

class SeoPage extends StatefulWidget {
  const SeoPage({super.key});

  @override
  State<SeoPage> createState() => _SeoPageState();
}

class _SeoPageState extends State<SeoPage> {
  @override
  Widget build(BuildContext context) {
    return SizedBox(
        height: MediaQuery.of(context).size.height,
        width: MediaQuery.of(context).size.width,
        child: Stack(
          alignment: Alignment.center,
          children: [
            const wrapp.Image(
              src: './assets/assets/wrapp.png',
              fit: BoxFit.cover,
            ),
            Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                wrapp.Text(
                  'Hello Seo!',
                  style: TextStyle(
                      fontFamily: 'arial',
                      fontSize: MediaQuery.of(context).size.width / 6,
                      color: Colors.white,
                      fontWeight: FontWeight.w900),
                  tag: wrapp.Tag.h1,
                ),
                const SizedBox(
                  height: 300,
                )
              ],
            ),
            Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Container(
                  height: MediaQuery.of(context).size.height / 2,
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    wrapp.Image(
                      src: './assets/assets/github.png',
                      fit: BoxFit.cover,
                      height: MediaQuery.of(context).size.width / 20,
                      width: MediaQuery.of(context).size.width / 20,
                    ),
                    const SizedBox(
                      width: 20,
                    ),
                    wrapp.Text(
                      'Download',
                      href: 'https://github.com/Antoinegtir',
                      style: TextStyle(
                          fontFamily: 'arial',
                          fontSize: MediaQuery.of(context).size.width / 25,
                          color: Colors.white,
                          fontWeight: FontWeight.w900),
                      tag: wrapp.Tag.h1,
                    ),
                  ],
                ),
                Container(
                  height: MediaQuery.of(context).size.height / 4,
                ),
                Row(
                  children: [
                    Container(
                      width: 20,
                    ),
                    wrapp.Image(
                      height: MediaQuery.of(context).size.width / 20,
                      width: MediaQuery.of(context).size.width / 20,
                      borderRadius: BorderRadius.circular(200),
                      href: 'https://github.com/Antoinegtir',
                      src:
                          'https://avatars.githubusercontent.com/u/114834504?v=4',
                      fit: BoxFit.cover,
                    ),
                    Container(
                      width: 20,
                    ),
                    wrapp.Text(
                      'By Antoine Gonthier',
                      href: 'https://github.com/Antoinegtir',
                      style: TextStyle(
                          fontFamily: 'arial',
                          fontSize: MediaQuery.of(context).size.width / 40,
                          color: Colors.white,
                          fontWeight: FontWeight.w900),
                      tag: wrapp.Tag.h1,
                    ),
                  ],
                )
              ],
            )
          ],
        ));
  }
}
